iceoryx_hoofs/source/posix_wrapper/signal_watcher.cpp

```diff
@@ -29,26 +29,18 @@ void internalSignalHandler(int) noexcept
     for (uint64_t remainingNumberOfWaiters = instance.m_numberOfWaiters.load(); remainingNumberOfWaiters > 0;
          --remainingNumberOfWaiters)
     {
-        instance.m_semaphore.post().or_else([](auto) {
-            constexpr const char MSG[] = "Unable to increment semaphore in signal handler";
-            auto result = write(STDERR_FILENO, MSG, sizeof(MSG));
-            IOX_DISCARD_RESULT(result);
-            std::abort();
-        });
+        instance.m_semaphore->post().expect("Unable to increment semaphore in signal handler");
     }
 }
 
 SignalWatcher::SignalWatcher() noexcept
-    : m_semaphore{std::move(Semaphore::create(CreateUnnamedSingleProcessSemaphore, 0U)
-                                .or_else([](auto) {
-                                    std::cerr << "Unable to create semaphore for signal watcher" << std::endl;
-                                    constexpr bool UNABLE_TO_CREATE_SEMAPHORE_FOR_SIGNAL_WATCHER = false;
-                                    cxx::Ensures(UNABLE_TO_CREATE_SEMAPHORE_FOR_SIGNAL_WATCHER);
-                                })
-                                .value())}
-    , m_sigTermGuard(registerSignalHandler(Signal::TERM, internalSignalHandler))
+    : m_sigTermGuard(registerSignalHandler(Signal::TERM, internalSignalHandler))
     , m_sigIntGuard(registerSignalHandler(Signal::INT, internalSignalHandler))
 {
+    UnnamedSemaphoreBuilder()
+        .isInterProcessCapable(false)
+        .create(m_semaphore)
+        .expect("Unable to create semaphore for signal watcher");
```

Comment:
This should be safe. Although `getInstance` will be called in the signal handler and `getInstance` creates the instance, it must be called before the signal handler is registered.

Out of curiosity. It seems the signal handler can only be registered with a call to `getInstance`, which happens when one calls e.g. `waitForTerminationRequest()`, which in turn is called quite late. What happens when a signal was raised before this function was called? I guess the same as when no signal handler is registered? Do we want to give the user the opportunity to explicitly register the signal handler quite early in the main method?

Reply:
@elfenpiff I think I didn't communicate my intention clearly. It is not about the user to register a signal handler but about the abstraction registering the signal handler quite late. Let's assume `Ctrl+C` is pressed before the first call to `iox::posix::hasTerminationRequested`. In this scenario there is no graceful termination with stack unwinding and all sockets, etc. are not cleaned up. If one wants to use the convenient `iox::posix::hasTerminationRequested` function but also terminate gracefully, there is no obvious way to do it, except calling `iox::posix::hasTerminationRequested` right at the start of `main`. The question is, do we want to give the user a more obvious way to register our signal handler at the start of main?

There is of course also a downside of this since there might be blocking calls the user added which are stuck but then a custom signal handler needs to be registered.

Reply:
This is a bigger question and I created an issue to discuss it in more detail and come up with a clean solution: #1414